File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/128-longest-consecutive-sequence-approach2.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function longestConsecutive(nums: number[]): number {
    if (nums.length === 0) return 0;                               // L1: O(1) guard
    const unique = [...new Set(nums)].sort((a, b) => a - b);     // L2: O(n log n) sort after O(n) dedup
    let best = 1, cur = 1;                                        // L3: O(1)
    for (let i = 1; i < unique.length; i++) {                    // L4: loop, n iterations
        if (unique[i] === unique[i - 1] + 1) {                   // L5: O(1) comparison
            cur++;
            best = Math.max(best, cur);                           // L7: O(1)
        } else {
            cur = 1;                                               // L8: O(1) reset
        }
    }
    return best;
}

assert(longestConsecutive([100, 4, 200, 1, 3, 2]) === 4);
assert(longestConsecutive([0, 3, 7, 2, 5, 8, 4, 6, 0, 1]) === 9);
assert(longestConsecutive([]) === 0);
assert(longestConsecutive([1]) === 1);
assert(longestConsecutive([1, 2, 3, 4, 5]) === 5);
assert(longestConsecutive([5, 4, 3, 2, 1]) === 5);
console.log("all tests pass");

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/128-longest-consecutive-sequence-approach2.py
def longest_consecutive(nums: list[int]) -> int:
    if not nums:
        return 0
    nums_sorted = sorted(set(nums))
    best = cur = 1
    for i in range(1, len(nums_sorted)):
        if nums_sorted[i] == nums_sorted[i - 1] + 1:
            cur += 1
            best = max(best, cur)
        else:
            cur = 1
    return best

assert longest_consecutive([100, 4, 200, 1, 3, 2]) == 4
assert longest_consecutive([0, 3, 7, 2, 5, 8, 4, 6, 0, 1]) == 9
assert longest_consecutive([]) == 0
assert longest_consecutive([1]) == 1
assert longest_consecutive([1, 2, 3, 4, 5]) == 5
assert longest_consecutive([5, 4, 3, 2, 1]) == 5
print("all tests pass")
